code/modules/mob/living/emote.dm

```diff
@@ -583,45 +583,88 @@
 /datum/emote/living/custom/can_run_emote(mob/user, status_check, intentional)
 	. = ..() && intentional
 
-/datum/emote/living/custom/proc/check_invalid(mob/user, input)
+/datum/emote/living/custom/proc/emote_is_valid(mob/user, input)
+	if(CAN_BYPASS_FILTER(user))
+		return TRUE
+
 	var/static/regex/stop_bad_mime = regex(@"says|exclaims|yells|asks")
 	if(stop_bad_mime.Find(input, 1, 1))
 		to_chat(user, span_danger("Invalid emote."))
-		return TRUE
-	return FALSE
+		return FALSE
+
+	var/list/filter_result = is_ic_filtered(input)
+
+	if(filter_result)
+		to_chat(user, span_warning("That emote contained a word prohibited in IC emotes! Consider reviewing the server rules."))
+		to_chat(user, span_warning("\"[input]\""))
+		REPORT_CHAT_FILTER_TO_USER(user, filter_result)
+		log_filter("IC Emote", input, filter_result)
+		SSblackbox.record_feedback("tally", "ic_blocked_words", 1, lowertext(config.ic_filter_regex.match))
+		return FALSE
+
+	filter_result = is_soft_ic_filtered(input)
+
+	if(filter_result)
+		if(tgui_alert(user,"Your emote contains \"[filter_result[CHAT_FILTER_INDEX_WORD]]\". \"[filter_result[CHAT_FILTER_INDEX_REASON]]\", Are you sure you want to emote it?", "Soft Blocked Word", list("Yes", "No")) != "Yes")
```

Comment:
How do these behave on clientless mobs, since we have automated emotes and these will run for them too

Reply:
I'm making a big assumption now that non-admins can't force clientless mobs to custom emote, and I added a quick short circuit that early returns if the emoting mob is clientless as a result. I can't think of any situations where it would be possible, at least.